pallets/accounts/src/lib.rs

```diff
@@ -385,16 +389,71 @@ pub mod pallet {
 		pub fn link(origin: OriginFor<T>, signature: ecdsa::Signature) -> DispatchResult {
 			let sender = ensure_signed(origin)?;
 
-			// the idea is to prove to this chain that the sender knows the private key of the external address
-			let mut message = b"EVM2Fragnova".to_vec();
-			message.extend_from_slice(&T::EthChainId::get().to_be_bytes());
-			message.extend_from_slice(&sender.encode());
+			let genesis_hash = <frame_system::Pallet<T>>::block_hash(T::BlockNumber::zero());
+			let genesis_hash_string = hex::encode(genesis_hash);
+
+			let sender_string = hex::encode(sender.encode());
+
+			// Metamask signTypedData_v4 - https://jsfiddle.net/4mwu2g80/43/
+			// We emulate JS JSON.stringify behavior here basically...
+			// let message = format!(r#"{{"domain":{{"name":"Fragnova Network","version":"1","chainId":{},"verifyingContract":"0xF5A0Af5a0AF5a0AF5a0af5A0Af5A0AF5a0AF5A0A"}},"message":{{"fragnovaGenesis":"{}","op":"link","sender":"{}"}},"primaryType":"Msg","types":{{"EIP712Domain":[{{"name":"name","type":"string"}},{{"name":"version","type":"string"}},{{"name":"chainId","type":"uint256"}},{{"name":"verifyingContract","type":"address"}}],"Msg":[{{"name":"fragnovaGenesis","type":"string"}},{{"name":"op","type":"string"}},{{"name":"sender","type":"string"}}]}}}}"#, T::EthChainId::get(), genesis_hash, sender_string);
+
+			// TODO, THE ABOVE IS WRONG.
+			// We need to follow https://eips.ethereum.org/EIPS/eip-712 spec.. which is convoluted...
+
+			// We encode the message using the following encoding function:
+			// encode(domainSeparator : 𝔹²⁵⁶, message : 𝕊) = "\x19\x01" ‖ domainSeparator ‖ hashStruct(message).
+			// See: https://eips.ethereum.org/EIPS/eip-712#specification-1
+			let message: Vec<u8> = [
```

Comment:
Love it but can you write a shared common method so it can be used around the projects and tests? Thanks

Reply:
> Love it but can you write a shared common method so it can be used around the projects and tests? Thanks

A common method that returns a link message, right?

Reply:
yep